### PR TITLE
DashCanvas bug fixes

### PR DESCRIPTION
**Pull request by ghsolomon**
Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [x] Added CHANGELOG entry, or determined not required.
- [x] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [x] Updated doc comments / prop-types, or determined not required.
- [x] Reviewed and tested on Mobile, or determined not required.
- [x] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

**Comment by ghsolomon** [comment]
#2959

File: desktop/cmp/dash/canvas/DashCanvas.js [hunk 1]
```diff
@@ -26,48 +29,58 @@ export const [DashCanvas, dashCanvas] = hoistCmp.withFactory({
...
-                ContextMenu.show(
-                    dashCanvasContextMenu({
-                        dashCanvasModel: model,
-                        clickPosition: {x, y}
-                    }),
-                    {left: x, top: y},
-                    null,
-                    XH.darkTheme
-                );
+                const {clientX, clientY, target} = e,
```
Comment (lbwexler) [inline desktop/cmp/dash/canvas/DashCanvas.js:43]: put the variables and computation you need for the computation within the if block.

File: desktop/cmp/dash/canvas/DashCanvasModel.js [hunk 2]
```diff
@@ -86,6 +91,8 @@ export class DashCanvasModel extends HoistModel {
      * @param {number[]} c.margin - Between items [x,y] in pixels
      * @param {number} c.maxRows - Maximum number of rows permitted for this container
      * @param {number[]} c.containerPadding - Padding inside the container [x, y] in pixels
+     * @param {string} [c.emptyText] - text to display when the container is empty
+     * @param {string} [c.addViewButtonText] - text to display on the add view button
```
Comment (lbwexler) [inline desktop/cmp/dash/canvas/DashCanvasModel.js:95]: This seems rather specific, in any case, might call it `addViewText`.  Might it be used elsewhere? Do we need it?

File: desktop/cmp/dash/canvas/DashCanvasModel.js [hunk 3]
```diff
@@ -162,6 +173,7 @@ export class DashCanvasModel extends HoistModel {
     loadState(state) {
         this.clear();
         state.forEach(state => this.addView(state.viewSpecId, state));
+        this._publishStateImmediately();
```
Comment (lbwexler) [inline desktop/cmp/dash/canvas/DashCanvasModel.js:176]: why do we need to round-trip this immediately?  looks suspicious

File: desktop/cmp/dash/canvas/DashCanvasViewModel.js [hunk 4]
```diff
@@ -18,6 +19,14 @@ import {createRef} from 'react';
 export class DashCanvasViewModel extends DashViewModel {
     /** @member {RefObject<DOMElement>} */
     ref = createRef();
+    /** @member {boolean} */
+    @observable hidePanelHeader;
+
+    constructor(cfg) {
+        super(cfg);
+        makeObservable(this);
+        this.hidePanelHeader = cfg.viewSpec.hidePanelHeader ?? false;
```
Comment (lbwexler) [inline desktop/cmp/dash/canvas/DashCanvasViewModel.js:28]: use !!cfg.viewSpec.hidePanelHeader to cast this to a boolean.

File: desktop/cmp/dash/canvas/DashCanvasModel.js (hunk 2 above)
Comment (ghsolomon) [reply desktop/cmp/dash/canvas/DashCanvasModel.js:95]: I agree.  Only question is how important is it to be consistent with the old DashContainer here?  (This is the member name and doc used in the old DashContainerModel)

File: desktop/cmp/dash/canvas/DashCanvasModel.js (hunk 3 above)
Comment (ghsolomon) [reply desktop/cmp/dash/canvas/DashCanvasModel.js:176]: Also agreed-- I'll try to come up with a better approach.  This is in response to this bug:

Initial state is not set on state member for DashCanvasModel until 1 second after model creation due to debounce

File: desktop/cmp/dash/canvas/impl/DashCanvasContextMenu.js [hunk 5]
```diff
@@ -75,10 +76,14 @@ function createAddMenuItems({dashCanvasModel, clickPosition}) {
...
-                        id,
-                        {layout: addPosition}
-                    )
+                    actionFn: () => {
+                        const viewModel = dashCanvasModel.addView(
+                            id,
+                            {layout: addPosition}
+                        );
+                        // Delay allows view to render before scrolling into view
+                        delay(() => viewModel.ref.current.scrollIntoView({behavior: 'smooth'}), 1);
```
Comment (lbwexler) [inline desktop/cmp/dash/canvas/impl/DashCanvasContextMenu.js:85]: perhaps add a method on DashCanvasModel of the form 'ensureVisible' can imagine this is generally useful.  also prefer `wait().then()` for this construct. A number of advantages: 
+ Putting the number first more readable and having the function last allows easier formatting. 
+ Also can easily  use promise changes, await and exception handling

File: desktop/cmp/dash/canvas/impl/DashCanvasView.js [hunk 6]
```diff
@@ -29,52 +29,55 @@ export const dashCanvasView = hoistCmp.factory({
     model: uses(DashCanvasViewModel, {publishMode: ModelPublishMode.LIMITED}),
 
     render({model, className}) {
-        const {viewSpec, viewState, containerModel, id, positionParams, title, ref} = model,
-            {extraMenuItems, contentLocked, renameLocked} = containerModel;
+        const {viewSpec, viewState, containerModel, id, positionParams, title, ref, hidePanelHeader} = model,
+            {extraMenuItems, contentLocked, renameLocked} = containerModel,
+            headerProps = hidePanelHeader ? {} : {
```
Comment (lbwexler) [inline desktop/cmp/dash/canvas/impl/DashCanvasView.js:34]: consider making a component for the popover here -- will be less of a wall of code, and also more stable -- right now if the observables used in any of your menu items change you need to re-render this entire component!

File: desktop/cmp/dash/canvas/DashCanvas.js [hunk 7]
```diff
@@ -76,4 +71,58 @@ DashCanvas.propTypes = {
...
+                    })
+                ]
+            })
+        });
+    }
+);
+
+const onContextMenu = (e, model) => {
+    const {classList} = e.target;
+    if (classList.contains('react-grid-layout') ||
```
Comment (haynesjm42) [inline desktop/cmp/dash/canvas/DashCanvas.js:109]: Could we just check if `e.target === model.ref.current` here?

File: desktop/cmp/dash/canvas/DashCanvasViewModel.js [hunk 8]
```diff
@@ -17,10 +18,29 @@ import {createRef} from 'react';
...
 
     get positionParams() {
         const {containerModel, id} = this;
         return containerModel.layout.find(view => view.i === id);
     }
+
+    /** Scrolls the DashCanvasView into view */
+    ensureVisible() {
+        const {ref} = this;
+        when(() => ref.current).then(() => {
```
Comment (haynesjm42) [inline desktop/cmp/dash/canvas/DashCanvasViewModel.js:42]: Can replace this direct usage of mobx `when` with something like this:
```
this.addReaction({
    when: () => ref.current,
    run: (el) =>el.scrollIntoView({behavior: 'smooth', block: 'nearest'});
})
```

File: desktop/cmp/dash/canvas/DashCanvasModel.js [hunk 9]
```diff
@@ -162,6 +167,13 @@ export class DashCanvasModel extends HoistModel {
     loadState(state) {
         this.clear();
         state.forEach(state => this.addView(state.viewSpecId, state));
+        wait().then(() =>
```
Comment (haynesjm42) [inline desktop/cmp/dash/canvas/DashCanvasModel.js:170]: Not sure about this, I think we are over-complicating things here - pretty sure all we need is this line after the `this.loadState()` call in the ctor:
```
this.state = this.buildState();
```
That at least resolves my bug about the `state` being undefined for 1 second after creating the DashCanvasModel with initial state (causing some undesired behavior in app logic for checking if the dash state has been modified).

File: desktop/cmp/dash/canvas/DashCanvasModel.js (hunk 2 above)
Comment (lbwexler) [reply desktop/cmp/dash/canvas/DashCanvasModel.js:95]: 👍🏻  -- Got it

**Pull request merged**